instrumentation/servlet/servlet-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/common/service/ServletAndFilterAdviceHelper.java

```diff
@@ -49,23 +46,28 @@
       return;
     }
 
-    AtomicBoolean responseHandled = new AtomicBoolean(false);
-    ServletAccessor<REQUEST, RESPONSE> accessor = tracer.getServletAccessor();
-
-    // In case of async servlets wait for the actual response to be ready
-    if (accessor.isRequestAsyncStarted(request)) {
-      try {
-        accessor.addRequestAsyncListener(
-            request, new TagSettingAsyncListener<>(tracer, responseHandled, context));
-      } catch (IllegalStateException e) {
-        // org.eclipse.jetty.server.Request may throw an exception here if request became
-        // finished after check above. We just ignore that exception and move on.
-      }
+    if (mustEndOnHandlerMethodExit(tracer, request)) {
+      tracer.end(context, response);
     }
+  }
 
-    // Check again in case the request finished before adding the listener.
-    if (!accessor.isRequestAsyncStarted(request) && responseHandled.compareAndSet(false, true)) {
-      tracer.end(context, response);
+  /**
+   * Helper method to determine whether the appserver handler/servlet service/servlet filter method
+   * that started a span must also end it, even if no error was detected. Extracted as a separate
+   * method to avoid duplicating the comments on the logic behind this choice.
+   */
+  public static <REQUEST, RESPONSE> boolean mustEndOnHandlerMethodExit(
+      ServletHttpServerTracer<REQUEST, RESPONSE> tracer, REQUEST request) {
+
+    if (tracer.isAsyncListenerAttached(request)) {
+      // This request is handled asynchronously and startAsync instrumentation has already attached
+      // the listener.
+      return false;
     }
+
+    // This means that startAsync was not called (assuming startAsync instrumentation works
```

Comment:
I guess now that you don't use `isAsyncStarted` any more the servlet engine correctness part isn't really relevant

Reply:
This is just a debugging hint that if for some reason async tests should fail in some new appserver configuration, then one thing to check would be to make sure the `startAsync` advice is triggered and doesn't throw.